### PR TITLE
Simplify Base32 and Base64 conversions

### PR DESCRIPTION
**Pull request by sipa**
Generalize `ConvertBits` a bit to also be usable for the existing Base32 and Base64 convertions (rather than just for Bech32).

### CODE REVIEW

**Comment by MeshCollider** [comment]
Impressive simplification, Concept ACK

**Comment by fanquake** [comment]
All Travis failures seem to be the same issue in signmessages.py
```
    self.run_test()
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/signmessages.py", line 24, in run_test
    assert(self.nodes[0].verifymessage(address, signature, message))
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/coverage.py", line 47, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/authproxy.py", line 138, in __call__
    raise JSONRPCException(response['error'])
test_framework.authproxy.JSONRPCException: Malformed base64 encoding (-5)
2017-11-07 23:50:01.876000 TestFramework (INFO): Stopping nodes
2017-11-07 23:50:01.877000 TestFramework.node0 (DEBUG): Stopping node
```

**Comment by sipa** [comment]
@fanquake Thanks, fixed.

File: src/utilstrencodings.cpp [hunk 1]
```diff
@@ -346,102 +227,30 @@ std::vector<unsigned char> DecodeBase32(const char* p, bool* pfInvalid)
...
-                 mode = 7;
-                 break;
-
-             case 7: // we have 3 bits, and keep 0
-                 vchRet.push_back((left<<5) | dec);
-                 mode = 0;
-                 break;
-         }
+    const char* e = p;
+    std::vector<uint8_t> val;
```
Comment (promag) [inline src/utilstrencodings.cpp]: Any reason to remove the reserve? Doesn't pay off?

File: src/utilstrencodings.cpp [hunk 2]
```diff
@@ -346,102 +227,30 @@ std::vector<unsigned char> DecodeBase32(const char* p, bool* pfInvalid)
...
+
+    const char* q = p;
+    while (valid && *p != 0) {
+        if (*p != '=') {
+            valid = false;
+            break;
         }
+        ++p;
+    }
+    valid = valid && (p - e) % 8 == 0 && p - q < 8;
```
Comment (promag) [inline src/utilstrencodings.cpp]: Not sure but isn't `(p - e) & 7 == 0` faster? Or does the compiler optimise this case?

Comment (sipa) [reply src/utilstrencodings.cpp]: Strength reduction will do that for you at -O1 and higher.

**Comment by TheBlueMatt** [comment]
utACK 16a01bf1a2a318effbbac4a22f1092fee1924dc8. Would be nice to add these folks to test_bitcoin_fuzzy.cpp!

**Comment by laanwj** [comment]
Needs rebase in src/utilstrencodings.cpp.

**Comment by sipa** [comment]
Rebased.

File: src/utilstrencodings.cpp [hunk 3]
```diff
@@ -193,68 +158,30 @@ std::vector<unsigned char> DecodeBase64(const char* p, bool* pfInvalid)
...
-
-             case 3: // we have 2 bits and get 6
-                 vchRet.push_back((left<<6) | dec);
-                 mode = 0;
-                 break;
-         }
+    const char* e = p;
+    std::vector<uint8_t> val;
+    while (*p != 0) {
+        int x = decode64_table[(unsigned char)*p];
```
Comment (laanwj) [inline src/utilstrencodings.cpp]: It seems a bit of a pity to allocate and fill an intermediate vector here, which is just the contents of the string mapped through `decode64_table`. Could we maybe wrap the iterator to weave this into ConvertBits?

**Comment by laanwj** [comment]
Needs rebase after #11372

**Comment by sipa** [comment]
Rebased.

File: src/utilstrencodings.cpp (hunk 3 above)
Comment (sipa) [reply src/utilstrencodings.cpp]: @laanwj I actually tried that, and it became a bit more complex than I wanted to do in this PR.

Comment (laanwj) [reply src/utilstrencodings.cpp]: Yes it seems quite annoying to do so in C++, iterators are tricky enough to make this seemingly trivial thing non-trivial :/

So as we keep it like this, we should heed @promag's solution and reserve, I think.

**Comment by sipa** [comment]
Added reserves in several places.

**Comment by laanwj** [comment]
> Added reserves in several places.

Thanks!

utACK b3ea8ccb7af475703b97246a2baf2e105d24d6f9

**Pull request merged**